kernel/thread.c

```diff
@@ -1126,9 +1130,27 @@ int k_thread_runtime_stats_all_get(k_thread_runtime_stats_t *stats)
 /* Timeout handler for *_thread_timeout() APIs */
 void z_thread_timeout(struct _timeout *timeout)
 {
+	k_spinlock_key_t  key;
+	struct k_spinlock *lock;
 	struct k_thread *thread = CONTAINER_OF(timeout,
 					       struct k_thread, base.timeout);
 
+	/*
+	 * Wait until the spin lock associated with the object upon which
+	 * the thread was pended is available. If there was no such object
+	 * the spin lock would be z_obj_lock and can be expected to be
+	 * available. This provides simple protection against timeout induced
+	 * wait queue changes for kernel objects that must operate on multiple
+	 * waiting items.
```

Comment:
Not sure this explanation is completely right.  The lock parameter to pend_curr() is essentially a condition variable, though it doesn't use the term.  The pend call atomically (!) releases the lock and puts the thread to sleep, so there's no intermediate state visible.  This change just makes the timeout processing part of the same paradigm: the timeout is guaranteed to see one of (1) the lock held and the thread still-unpended or (2) the lock available and the thread asleep.  In the case of (1), taking the lock then guarantees that we wait until the thread is pended.